File: ipsc-backend/src/scoring.ts
export type PowerFactor = 'minor' | 'major';

const SCORING = {
  minor: { A: 5, C: 3, D: 1 },
  major: { A: 5, C: 4, D: 2 },
} as const;

function round(value: number, decimals = 4): number {
  return Math.round(value * 10 ** decimals) / 10 ** decimals;
}

export interface HitCounts {
  A: number;
  C: number;
  D: number;
  M: number;
  N: number;
}

export interface Penalties {
  PE: number;
}

export interface ScoreResult {
  totalPoints: number;
  hitFactor: number;
}

export function calculateScore(
  hits: HitCounts,
  penalties: Penalties,
  totalTime: number,
  powerFactor: PowerFactor
): ScoreResult {
  const pf = SCORING[powerFactor];
  const points = hits.A * pf.A + hits.C * pf.C + hits.D * pf.D;
  const penaltyPoints = (hits.M + hits.N) * 10 + penalties.PE * 10;
  const totalPoints = Math.max(points - penaltyPoints, 0);
  const hitFactor = totalTime > 0 ? totalPoints / totalTime : 0;
  return { totalPoints: round(totalPoints), hitFactor: round(hitFactor) };
}

File: ipsc-sim-mobile/sim-core/src/Scoring.cs
namespace IpscSim.Core;

/// <summary>
/// IPSC Comstock scoring: points per hit zone, depending on division power factor.
/// Misses and no-shoots = -10. Procedural / FTN = -10 (applied elsewhere).
/// </summary>
public static class Scoring
{
    public const int MissPenalty = -10;
    public const int NoShootPenalty = -10;
    public const int FtnPenalty = -10; // per missing required hit

    public static int PointsFor(HitZone zone, PowerFactor pf) => (zone, pf) switch
    {
        (HitZone.A, _) => 5,
        (HitZone.C, PowerFactor.Major) => 4,
        (HitZone.C, PowerFactor.Minor) => 3,
        (HitZone.D, PowerFactor.Major) => 2,
        (HitZone.D, PowerFactor.Minor) => 1,
        (HitZone.Miss, _) => MissPenalty,
        (HitZone.NoShoot, _) => NoShootPenalty,
        _ => 0
    };

    /// <summary>HF = points / time, clamped non-negative, 0 if time is 0.</summary>
    public static double HitFactor(int totalPoints, double totalTime)
    {
        if (totalTime <= 0) return 0;
        return Math.Max(0, totalPoints) / totalTime;
    }
}
